src/decompose_mesh_SCOTCH/scotch_5.1.11/man/man1/acpl.1
." Text automatically generated by txt2man
.TH acpl 1 "August 03, 2010" "" "Scotch user's manual"
.SH NAME
\fBacpl \fP- compile a target architectures
\fB
.SH SYNOPSIS
.nf
.fam C
\fBacpl\fP [\fIoptions\fP] [\fIitfile\fP] [\fIotfile\fP]
.fam T
.fi
.SH DESCRIPTION
The \fBacpl\fP program compiles a decomposition-defined target
architecture file \fIitfile\fP of type '\fBdeco\fP 0' into a compiled
decomposition-defined target architecture of type '\fBdeco\fP 1', stored
in file \fIotfile\fP.
.PP
Compiling a decomposition-defined architecture amounts to computing
the distance matrix of all possible subdomains, from the distance
matrix of all terminal subdomains provided in the '\fBdeco\fP 0' format.
Since this computation is internally performed every time a '\fBdeco\fP 0'
format is read, and is quadratic in time, pre-compiling the target
architecture by means of \fBacpl\fP can save some time when repeatedly
computing mappings on a large decomposition-defined '\fBdeco\fP 0' target
architecture.
.PP
When the proper libraries have been included at compile time, \fBacpl\fP
can directly handle compressed files, both as input and output. A
stream is treated as compressed whenever its name is postfixed with
a compressed file extension, such as in 'brol.tgt.bz2' or '-.gz'. The
compression formats which can be supported are the bzip2 format
('.bz2'), the gzip format ('.gz'), and the lzma format ('.lzma', on
input only).
.PP
Since decomposition-defined target architecture files have a size
which is quadratic in the number of target vertices, because of the
distance matrix structures, using compressed files to store them may
save a lot of space, especially for compiled target architecture
files.
.SH OPTIONS
.TP
.B
\fB-h\fP
Display some help.
.TP
.B
\fB-V\fP
Display program version and copyright.
.SH EXAMPLE
Create a compiled cube-connected-cycle target architecture
of dimension 4, and save it under the \fBgzip\fP(1) format to
file 'ccc4c.tgt.gz'. The dash '-' standard file name is
used so that the '\fBdeco\fP 0' target architecture description
produced by \fBamk_ccc\fP(1) is read from the standard input,
through the pipe.
.PP
.nf
.fam C
    $ amk_ccc 4 | acpl - ccc4c.tgt.gz

.fam T
.fi
.SH SEE ALSO
\fBamk_grf\fP(1), \fBamk_ccc\fP(1), \fBatst\fP(1), \fBdgmap\fP(1), \fBgmap\fP(1), \fBgmtst\fP(1).
.PP
Scotch user's manual.
.SH AUTHOR
Francois Pellegrini <francois.pellegrini@labri.fr>


src/decompose_mesh_SCOTCH/scotch_5.1.10b/src/scotch/acpl.c
/* Copyright 2004,2007,2008,2010 ENSEIRB, INRIA & CNRS
**
** This file is part of the Scotch software package for static mapping,
** graph partitioning and sparse matrix ordering.
**
** This software is governed by the CeCILL-C license under French law
** and abiding by the rules of distribution of free software. You can
** use, modify and/or redistribute the software under the terms of the
** CeCILL-C license as circulated by CEA, CNRS and INRIA at the following
** URL: "http://www.cecill.info".
** 
** As a counterpart to the access to the source code and rights to copy,
** modify and redistribute granted by the license, users are provided
** only with a limited warranty and the software's author, the holder of
** the economic rights, and the successive licensors have only limited
** liability.
** 
** In this respect, the user's attention is drawn to the risks associated
** with loading, using, modifying and/or developing or reproducing the
** software by the user in light of its specific status of free software,
** that may mean that it is complicated to manipulate, and that also
** therefore means that it is reserved for developers and experienced
** professionals having in-depth computer knowledge. Users are therefore
** encouraged to load and test the software's suitability as regards
** their requirements in conditions enabling the security of their
** systems and/or data to be ensured and, more generally, to use and
** operate it in the same conditions as regards security.
** 
** The fact that you are presently reading this means that you have had
** knowledge of the CeCILL-C license and that you accept its terms.
*/
/************************************************************/
/**                                                        **/
/**   NAME       : acpl.c                                  **/
/**                                                        **/
/**   AUTHOR     : Francois PELLEGRINI                     **/
/**                                                        **/
/**   FUNCTION   : This module compiles target architec-   **/
/**                ture decomposition files for fast       **/
/**                loading.                                **/
/**                                                        **/
/**   DATES      : # Version 2.0  : from : 12 nov 1994     **/
/**                                 to   : 12 nov 1994     **/
/**                # Version 3.0  : from : 06 jul 1995     **/
/**                                 to   : 19 sep 1995     **/
/**                # Version 3.2  : from : 24 sep 1996     **/
/**                                 to   : 12 may 1998     **/
/**                # Version 3.3  : from : 02 oct 1998     **/
/**                                 to   : 02 oct 1998     **/
/**                # Version 3.4  : from : 03 feb 2000     **/
/**                                 to   : 03 feb 2000     **/
/**                # Version 4.0  : from : 09 feb 2004     **/
/**                                 to   : 09 feb 2004     **/
/**                # Version 5.0  : from : 23 dec 2007     **/
/**                                 to   : 16 mar 2008     **/
/**                # Version 5.1  : from : 01 jul 2010     **/
/**                                 to   : 01 jul 2010     **/
/**                                                        **/
/************************************************************/

/*
**  The defines and includes.
*/

#define ACPL

#include "module.h"
#include "common.h"
#include "scotch.h"
#include "acpl.h"

/*
**  The static variables.
*/

static int                  C_fileNum = 0;        /* Number of file in arg list */
static File                 C_fileTab[C_FILENBR] = { /* File array              */
                              { "-", NULL, "r" },
                              { "-", NULL, "w" } };

static const char *         C_usageList[] = {
  "acpl [<input target file> [<output target file>]] <options>",
  "  -h  : Display this help",
  "  -V  : Print program version and copyright",
  NULL };

/******************************/
/*                            */
/* This is the main function. */
/*                            */
/******************************/

int
main (
int                         argc,
char *                      argv[])
{
  SCOTCH_Arch         arch;                       /* Architecture read and written */
  int                 i;

  errorProg ("acpl");

  if ((argc >= 2) && (argv[1][0] == '?')) {       /* If need for help */
    usagePrint (stdout, C_usageList);
    return     (0);
  }

  for (i = 0; i < C_FILENBR; i ++)                /* Set default stream pointers */
    C_fileTab[i].pntr = (C_fileTab[i].mode[0] == 'r') ? stdin : stdout;
  for (i = 1; i < argc; i ++) {                   /* Loop for all option codes                        */
    if ((argv[i][0] != '-') || (argv[i][1] == '\0') || (argv[i][1] == '.')) { /* If found a file name */
      if (C_fileNum < C_FILEARGNBR)               /* File name has been given                         */
        C_fileTab[C_fileNum ++].name = argv[i];
      else {
        errorPrint ("main: too many file names given");
        return     (1);
      }
    }
    else {                                        /* If found an option name */
      switch (argv[i][1]) {
        case 'H' :                                /* Give the usage message */
        case 'h' :
          usagePrint (stdout, C_usageList);
          return     (0);
        case 'V' :
          fprintf (stderr, "acpl, version " SCOTCH_VERSION_STRING "\n");
          fprintf (stderr, "Copyright 2004,2007,2008,2010 ENSEIRB, INRIA & CNRS, France\n");
          fprintf (stderr, "This software is libre/free software under CeCILL-C -- see the user's manual for more information\n");
          return  (0);
        default :
          errorPrint ("main: unprocessed option (\"%s\")", argv[i]);
          return     (1);
      }
    }
  }

  fileBlockOpen (C_fileTab, C_FILENBR);           /* Open all files */

  SCOTCH_archInit (&arch);                        /* Initialize architecture structure */
  SCOTCH_archLoad (&arch, C_filepntrtgtinp);      /* Load architecture                 */
  if (strcmp (SCOTCH_archName (&arch), "deco") != 0) { /* If not a decomposition       */
    errorPrint ("main: architecture is not decomposition-defined");
    return     (1);
  }
  SCOTCH_archSave (&arch, C_filepntrtgtout);      /* Save the compiled architecture */

  fileBlockClose (C_fileTab, C_FILENBR);          /* Always close explicitely to end eventual (un)compression tasks */

  SCOTCH_archExit (&arch);

#ifdef COMMON_PTHREAD
  pthread_exit ((void *) 0);                      /* Allow potential (un)compression tasks to complete */
#endif /* COMMON_PTHREAD */
  return (0);
}
